### DISCUSSION TITLE
KafkaScaler - How to get rid of  / How to handle "invalid offset found for topic" warnings after Topics have been recreated in Kafka for DEV ?

### DISCUSSION THREAD

**Discussion by QuintinBecker** [kedacore/keda#1191] [General]
We are using KEDA in OKD as central deployment and use it for DEV and PROD branches at the same time. 
Same is true for our Kafka which is used for all branches.
As we build new pipelines and reset Kafka-Topics during this proces we need to reset the stored metrics in Keda as well for this topics. 

Is there a planed procedure to handling the reset of a topic that is used by scaledObjects ? 

In addition we see aome of the deployments not scaling to zero in this state and assume this is linked to this problem. 

Environment:
OKD 4.5 
Keda 2.0
Rollout is handled by Argo v1.7.6+b04c25e

**Comment by zroubalik** [comment]
> we need to reset the stored metrics in Keda as well for this topics

What do you mean by this exactly?

**Reply by QuintinBecker** [reply]
For me it seems as some metrics have been stored within the Keda-metrics. If we now delete a topic and redeploy the pipeline we have offset of 0 in Kafka and I assume the old offset is still stored in Keda-metrics, which leads to this message. We use the same topic names and consumergroups after redeploying.

I can check the topics in Kafka after redeploying the pipeline and can confirm that offset is reset and increasing as we start processing date in our pipeline. 

Maybe I am totally wrong.

**Reply by zroubalik** [reply]
KEDA doesn't persist any metrics, it is always asking Kafka for the current state. I am not Kafka expert, @grassiale could chime in?

**Reply by grassiale** [reply]
Hi @QuintinBecker there is no concept of persistence or history of lag in keda.
Right now, when a new consumer with no offsets committed is created, keda behaves accordingly to the offsetResetPolicy specified in the configuration.
If the policy is set to earliest (a new consumer wants to replay everything in the topic from its beginning) and no offset is committed, the scaler will return a lag value equal to the last offset of the topic ( in the case of a new topic 0), so it should scale your deployment to 0 replicas.
If the policy is set to latest (so the new consumer will only consume new messages) and no offset is committed, the scaler will return a negative lag value, and will also tell the hpa to remain active, hence the deployment should have one replica running, so that the consumer will be able to consume new messages, and commit its offset.

So maybe you are incurring in the second case? Please let me know if this sounds right to you.

**Reply by zroubalik** [reply]
> Right now, when a new consumer with no offsets committed is created, keda behaves accordingly to the offsetResetPolicy specified in the configuration.
> If the policy is set to earliest (a new consumer wants to replay everything in the topic from its beginning) and no offset is committed, the scaler will return a lag value equal to the last offset of the topic ( in the case of a new topic 0), so it should scale your deployment to 0 replicas.
> If the policy is set to latest (so the new consumer will only consume new messages) and no offset is committed, the scaler will return a negative lag value, and will also tell the hpa to remain active, hence the deployment should have one replica running, so that the consumer will be able to consume new messages, and commit its offset.
> 

@grassiale thanks for jumping in! 

BTW It would be really great if you can add these kind of details on the offset policy to the [docs](https://keda.sh/docs/2.0/scalers/apache-kafka/), once you have some time :) Thanks!

**Reply by QuintinBecker** [reply]
Sorry for my late response, we made some tests and chenged some points on the pipeline. 
Thank you for all the info on the scaling process. 

In our case we destruct the pipeline (scaled objects and deployments) and all topics linkt to it and redeploy everything. In this state the offset is default -1001 or so and there should be no scaling of the deployments till we get some events running through the pipeline. 

As we have 5 partitions we get assignment of a single consumer to all of them:

```
[PYTHON][INFO][2020-10-06 07:11:58,962]  Error Message in Kafka: TOPIC_1 1 0
[PYTHON][INFO][2020-10-06 07:11:58,962]  % TOPIC_1 [1] reached end at offset 0

[PYTHON][INFO][2020-10-06 07:11:58,962]  Error Message in Kafka: TOPIC_1 2 0
[PYTHON][INFO][2020-10-06 07:11:58,962]  % TOPIC_1 [2] reached end at offset 0

[PYTHON][INFO][2020-10-06 07:11:58,962]  Error Message in Kafka: TOPIC_1 0 0
[PYTHON][INFO][2020-10-06 07:11:58,962]  % TOPIC_1 [0] reached end at offset 0

[PYTHON][INFO][2020-10-06 07:11:59,462]  Error Message in Kafka: TOPIC_1 4 0
[PYTHON][INFO][2020-10-06 07:11:59,462]  % TOPIC_1 [4] reached end at offset 0

[PYTHON][INFO][2020-10-06 07:11:59,464]  Error Message in Kafka: TOPIC_1 3 0
[PYTHON][INFO][2020-10-06 07:11:59,464]  % TOPIC_1 [3] reached end at offset 0
```

After that the Deployment is waiting and is not scaled down to 0. 

Do we have to initialize an offset to all partitions to get rid of this behaviour ?

**Reply by grassiale** [reply]
HI @QuintinBecker your assumption is correct. Until all partitions won't have a committed offset, keda will not scale to 0. 
I think this behaviour can be changed, but I still haven't put my mind into it.

**Reply by QuintinBecker** [reply]
Hi @grassiale thanks for the feedback.
It is not a big deal to get all partitions "initialized" on our side now that we know it is not an issue per se. 
Do you plan to adopt the behavior or do you prefer a workaround on the customer side (catch offsets of -1001 and set offset to 0 by client) ?
Can you maybe have a look on my request #1190 as well ? This would make life much easier for us.

**Reply by iferca** [reply]
@QuintinBecker I'm having similar "issue", just deploying my ScaledJob I start getting this message: 

`invalid offset found for topic topic-test in group cg-test and partition 0, probably no offset is committed yet`

and I don't see any scaling happening even if I post something in the topic while I can still consume from the terminal using kafka cli. 

In your previous post you said _"It is not a big deal to get all partitions "initialized" on our side"_ abd I undertand this solved your issue. Can you please enlight me about how did you initialized the partions, what do you mean by that?

**Reply by QuintinBecker** [reply]
Hi @iferca,
we had an issue after we had removed the topics (and all events) from Kafka and redeployed the same named topic. This is done for testing our event-driven pipeline. 
We got this message because we had removed all events and there was no offset set after the topic had been created. This situation led to the deployment not scaling down to zero. So we had the deployment running even if ther has been no lag on the topic and consumergroup. In addition we are running the consumer with setting autocommit = false and have to handle each event seperately.
"Initialization" in this case means simply to create at least one Event in each patition of the topic and let the consumer set the offset correct. 

If I get you right your deployments are not scaling properly even if there is lag for the consumergroup and topic. 
Common Issues we have seen in the past were a mismatch of the conumsergroups in the deployment and its scaled object.
In addition keep in mind that an error within the deployment can lead to this behaviour as well. So please check the deployment itself and try to deploy it by hand.
